rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,270 @@
+
+
+## Migrating from distributed to centeralised (OCM Hub)
+
+This is not covered.
+
+## Migrating from single to multiple cluster (distributed)
+
+The steps for this are quite straight forward:
+This assumes you already have 1 cluster with a gateway and DNSPolicy setup.
+- Create a new gateway in a second cluster that has a listener defined with a hostname you want to use across clusters
+- Create a DNSPolicy that uses the same strategy as the DNSPolicy already configured that targets this gateway (very likely a mirror copy)
+- Ensure you create and link to the same DNS Provider (via credential) with the existing zone
+
+## Cleanup
+
+Cleanup of a cluster's zone file entries is triggered by deleting the DNS Policy that caused the records to be created. The DNSPolicy will have a finalizer, when it is deleted the Kuadrant operator will mark the created DNSRecord for deletion. The DNS record will have a finalizer from the DNS Operator which will clean the records from the zone before removing the finalizer.
+
+## Switching strategy (IE from LoadBalanced to Simple or the other way)
+The DNSPolicy offers 2 strategies 
+1) Simple (one A record multiple values)
+2) LoadBalanced (leverages CNAMES and A records to provide advanced DNS based routing to be used (example GEO and Weighted) )
+
+The strategy field will be immutable once set. To migrate from one strategy to the other you must first delete the existing policy and recreate it with a new strategy. In a multi-cluster environment, this can mean you have a period of time where 2 different strategies are set (simple on one and load balanced on another). The controller will mitigate against this in the following way. When attempting to update the remote zone, it first does a read. If it spots existing records in that zone that correlate with a different strategy, it will not write its own records but will flag this as an error state in the status. So the records will remain in the original strategy until all policies for that listener host match (more in this below).
+
+
+## Orphan Mitigation
+
+It is possible for an orphan set of records to be created if for example a cluster is removed and the controller on that cluster is not given time to clean up. This is not something we currently mitigate against directly in the controller. However, you can define health checks that would remove these orphans from the DNS response. Additionally, there is future work that leverages a "heart beat" that should allow each alive controller to see and remove a dead controllers records within a given criteria.
```

Comment:
`However, you can define health checks that would remove these orphans from the DNS response.` -> In the end this may lead to serving of orphaned ip instead of NXDOMAIN if all of the other DNSPolicies from other clusters are removed (or how does the dns provider handle situations when there are no healthy endpoints left?).

Reply:
it returns all IPs if no healthy IPs left

Reply:
ok, then if the orphan is the latest record it gets served (instead of NXDOMAIN) despite health checks, right?

Reply:
another fact that arises from not-mitigating orphans is that the strategy can't be changed for this hostname if orphan is not manually removed. (actually this may be a feature)